[12.0][IMP] - Make recurrence mechanism on contract line

sbejaoui opened the pull request:
Make recurrence mechanism on contract line and some other refactoring
Related to #200

sbejaoui committed:
[IMP] - Make recurrence mechanism on contract line

Make recurrence mechanism on contract line and some other refactoring

[FIX] - Keep contract_cron on account_analytic_account model

contract_cron defined with no_update option.
Changing it, will cause issue to past version installation.

[IMP] - Fix recurring_next_date default value

recurring_next_date should have start_date as default value in prepaid policy
and start_date + invoicing_interval if postpaid

[FIX] - Fix test check no journal

[IMP] - Return created invoices on recurring_create_invoice

[IMP] - Specific process to compute recurring_next_date for  monthly-last-day

fixes: #198

[ADD] - Add Post-migration script to bring recurrence info from contract to contract lines

[ADD] - Add search filter based on date_end and recurring_next_date

 - not_finished filter in contract search view
 - finished filter in contract search view
 - Next Invoice group by in contract search view

[ADD] - Add unit tests

- cases to compute first recurring next date
- contract recurring_next_date
- contract date_end

[IMP] - Improve Unit tests

sbejaoui commented:
@gva-acsone, it's ready for a shot of test.

gva-acsone commented:
@sbejaoui , I've an error message when trying to add an invoice line on my contract template.
Can you check please? 
...
  File "src/lxml/parser.pxi", line 706, in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:106810)
  File "src/lxml/parser.pxi", line 646, in lxml.etree._raiseParseError (src/lxml/lxml.etree.c:105956)
  File "<string>", line 0
lxml.etree.XMLSyntaxError: <no detail available>

gva-acsone commented:
1. On the contract lines, I would switch the end date and the next invoice date so that the end date is located just next to the start date.

2. I managed to create my first invoice, but afterwards, Odoo refuses to create my next invoices.

sbejaoui commented:
@gva-acsone 
> 1. On the contract lines, I would switch the end date and the next invoice date so that the end date is located just next to the start date.

I will make the change for this

> 2. I managed to create my first invoice, but afterwards, Odoo refuses to create my next invoices.

The next date invoice > the today date, that's why, odoo would not create a new invoice. Nothing to invoice until the next date invoice.

gva-acsone commented:
Can we modify this so that we can create invoices in advance?

gva-acsone reviewed (approved):
👍 Functional test

pedrobaeza commented on contract/data/contract_cron.xml:
```diff
@@ -3,9 +3,9 @@
     <record model="ir.cron" id="account_analytic_cron_for_invoice">
         <field name="name">Generate Recurring Invoices from Contracts</field>
-        <field name="model_id" ref="analytic.model_account_analytic_account"/>
+        <field name="model_id" ref="contract.model_account_analytic_invoice_line"/>
```
This doesn't serve for existing installations as this won't be overwritten being noupdate=1

pedrobaeza replied:
Any way, why moving this cron to lines? We can keep the same and call from that one whatever its' needed.

sbejaoui committed:
[12.0][IMP] - Add strat/stop wizard to contract line

[12.0][IMP] - Add pause button to contract line

[IMP] - Add state filed in contract line form

[FIX] - stop don't change date_end for finished contract line

[IMP] - Change contract line buttons visibility

Add renewal process with termination notice

[FIX] - don't consider stop_date If it is after the contract line end_date

[IMP] - consider more cases in stop_plan_successor

[IMP] - cancel upcoming line on stop

[IMP] - Chnage next invoice date on un-cancel

[IMP] - Post message in contract on contract line actions

[IMP] - check contract line overlap

rafaelbn reviewed (commented):
:scream: :scream: :scream: 

Are you going to make main contract module line by line?

rafaelbn reviewed (changes requested):
Please let me know some arguments. I disagree.

- Customers don't one to define dates or invoicing types line by line
- If you create a invoice by error and after you deleted then you have to change dates for every line and more more cases

pedrobaeza commented:
I think recurrence by line is a good move, but with some UX refinements (I have not tried yet the PR):

* A security group for hiding recurrence by line by default.
* Recurrence definition in header, propagated to lines when set (silently if the security group is not present, and hidden on contrary).
* The same for next invoice date.

The rest can continue to happen at line level.

rafaelbn commented:
I guess that if @sbejaoui and @acesone in general were using this module like us from several years and their customersss were using it then the approach would be different, because you cannot change the functionality and UX so hard. Take thousands of  contract with several lines each to define all line by line.

In this PR we are changing the module for a **1 single customer case** but we have fit the standard contract module to dozens of customer. And I suppose that more Odoo integrators will be damaged by this if we don't take care of the situation.

pedrobaeza commented:
@rafaelbn please check my comment for the suggestion for having everything the same way but also with the possibility of defining recurrence by lines. I'm now thinking that maybe instead the security group, we can put a check at contract level for deciding if doing recurrence at line level.

sbidoul commented:
@rafaelbn I understand your concerns.

It's not only for one customer: the approach here is much more generic and enables many interesting features for contracts with many lines (such as having a subscriptions with some options that start and stop during the life of the contract). Also possibilities to add lines to existing contracts from the sale order, etc.

Migrating the data from previous version will work seamlessly.

Now indeed, from a UI perspective, if you have use cases with contracts with many lines where it is important to keep the recurrence information at the contract level, we can add something at the UI level to facilitate that. A checkbox on the contract saying "same recurrence on all lines" should be doable, with the recurrence field on contract being compute with inverse.

rafaelbn commented:
OK, thank you. I cannot test this again this week but next week.

rafaelbn reviewed (commented):
I will try my best review soon. Thank you

sbejaoui committed:
[FIX] - invoice last period for post-paid case

sbejaoui committed:
[IMP] - Add primary views for contract

[IMP] - don't use related filed for partner_id and pricelist_id

sbejaoui committed:
[IMP] - Adapt contract sale to contract new views

sbejaoui committed:
[FIX] - fix stop_plan_successor case 5

contract line start in the suspension period and end after it

sbejaoui committed:
[IMP] - improve cancel/uncancel process

[FIX] - Test if start_date is set before compute

[FIX] - date_end include in the period in auto_renew case

[FIX] - in suspension case, contract line should start a day after the end

[IMP] - confirm message on contract line cancel

rafaelbn commented:
Hello,

OK I agree @pedrobaeza 

Thank you @sbidoul for your comments.

Review

- [ ] I still don't see any way to define for all the lines:
  - Repeat every (number an interval)
  - Invoicing type
  - Date start
  - Date of next invoice
  - Date end is nor editable
- [x] The bif Title "Recurring invoices" is missing
- [x] The checkbox "Generate recurring invoices automatically" is missing 
- [x] The link "create invoices" is missing
- [x] The button Cost/Revenue is missing
- [ ] How to disable discounts? In v11 where disabled by default (so no need to use them)
- [ ] Where are the Tags?
- [ ] IMHO the Tab "other information" is not needed there aren't so many fields

sbidoul commented:
@rafaelbn a couple of comments of my side. (@sbejaoui will say more)

We will add the UX feature to manage recurrence globally at the contract level a bit later, don't worry.

About the Tags, and Cost/revenue buttons: it follows from a discussion with @pedrobaeza where we agreed to make separate views for contract that are independent from analytic account views. This is a first step towards splitting contracts from analytic accounts (we will not do the full split now, but we prepare for it). So the tags and cost/revenue buttons can be seen on the native analytic account view only. I think this is indeed better so people managing contracts (usually sales persons) do not see nor need to care about analytic accounting concepts. What would be nice though, is add a button to navigate to the analytic account from the contract form.

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+ALLOWED = namedtuple(
+    'ALLOWED',
+    ['PLAN_SUCCESSOR', 'STOP_PLAN_SUCCESSOR', 'STOP', 'CANCEL', 'UN_CANCEL'],
+)
```
This is a detail, but the python convention for named tuples, is CamelCase for the named tuple, and lower_case for the fields (since this is more or less a class with it's fields).

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+    canceled = is_canceled
+    return CRITERIA(
+        WHEN=when,
+        HAS_DATE_END=has_date_end,
```
```suggestion
        HAS_DATE_END=bool(date_end),
```

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+        WHEN=when,
+        HAS_DATE_END=has_date_end,
+        IS_AUTO_RENEW=is_auto_renew,
+        HAS_SUCCESSOR=has_successor,
```
```suggestion
        HAS_SUCCESSOR=bool(successor_contract_line_id),
```

sbejaoui commented:
Hello @rafaelbn,

Thank you for the feedback.
> * [ ]  I still don't see any way to define for all the lines:

Work in progress
> * [ ]  The bif Title "Recurring invoices" is missing
> * [ ]  The checkbox "Generate recurring invoices automatically" is missing

This checkbox was used in V8 as analytic accounts and contracts were in the same menu and have the same views. 
But, as the two concepts were separated since OCA migrated the module to V9, a contract has no use if this checkbox is unchecked.

> * [ ]  The link "create invoices" is missing
I replaced it by a native button on the form view visible in debug mode.
> * [ ]  How to disable discounts? In v11 where disabled by default (so no need to use them)
This was fixed at V12 migration, discount field is visible for users of `sale.group_discount_per_so_line`

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+        STOP=False,
+        CANCEL=False,
+        UN_CANCEL=False,
+    ),
```
I suggest removing these criteria with `None` values: this makes the logic a bit confusing.

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+                HAS_SUCCESSOR=None,
+                PREDECESSOR_HAS_SUCCESSOR=True,
+                CANCELED=True,
+            )
```
As mentioned above I think it's simpler to simply get the canceled case out of the matrix.
So something like `if is_canceled: return Allowed(False, False, False, ..., uncancel=some condition based on successor and predecssor)`?

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+        'HAS_DATE_END',
+        'IS_AUTO_RENEW',
+        'HAS_SUCCESSOR',
+        'PREDECESSOR_HAS_SUCCESSOR',
```
Why is this named PREDECESSOR_HAS_SUCCESSOR and not HAS_SUCCESSOR?

sbejaoui replied:
At cancel action we cut relation between the contract line and his predecessor, but in one way.
Only the successor of the predecessor is set to false. 
Like that, we permit again to perform plan_successor on the predecessor line.
We allow uncancel, only when the predecessor still without successor and lines are re-linked again.

That's why, we need `PREDECESSOR_HAS_SUCCESSOR` criteria for uncancel case.

sbidoul replied:
Ok, a short comment explaining that would help.

sbejaoui committed:
[IMP] - hide recurring_invoicing_type if recurring_rule_type is monthlylastday

for the monthlylastday case, pre-paid is logicly impossible,
if monthlylastday is set, we consider only post-paid case

[IMP] - Improve unit tests

sbidoul commented on contract/models/abstract_contract.py:
```diff
@@ -0,0 +1,69 @@
+
+class AbstractAccountAnalyticContract(models.AbstractModel):
+    _name = 'account.abstract.analytic.contract'
+    _description = 'Abstract Account Analytic Contract'
```
Let's use the descriptions to provide better names (while we keep the old model names for compatibility and ease of migration).

```suggestion
    _description = 'Abstract Recurring Contract'
```

sbidoul commented on contract/models/abstract_contract.py:
```diff
@@ -0,0 +1,69 @@
+        'res.company',
+        string='Company',
+        required=True,
+        default=lambda self: self.env.user.company_id,
```
`self.env['res.company']._company_default_get(self._name)`

sbidoul commented on contract/models/abstract_contract_line.py:
```diff
@@ -0,0 +1,204 @@
+
+class AccountAbstractAnalyticContractLine(models.AbstractModel):
+    _name = 'account.abstract.analytic.contract.line'
+    _description = 'Account Abstract Analytic Contract Line'
```
```suggestion
    _description = 'Abstract Recurring Contract Line'
```

sbidoul commented on contract/data/contract_line_constraints.py:
```diff
@@ -0,0 +1,281 @@
+# Copyright 2018 ACSONE SA/NV.
```
I would not put this python file in the data dir, this is surprising. Pls move it to models.

sbidoul commented on contract/models/account_invoice.py:
```diff
@@ -8,5 +8,5 @@ class AccountInvoice(models.Model):
-        'account.analytic.account',
-        string='Contract')
+        'account.analytic.account', string='Contract'
+    )
```
I still believe this should be removed.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        ondelete='cascade',
+        oldname='analytic_account_id',
+    )
+    date_start = fields.Date(string='Date Start', default=fields.Date.today())
```
should date_start be a required field?

sbidoul replied:
Nope, I forgot a contract is still an analytic account too, so not necessarily recurring.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        required=False,
+        readonly=True,
+        copy=False,
+        help="Contract Line created by this one.",
```
Elaborate a little bit the help text to explain in which circumstances a successor is created (suspension and restart, renewal, ...)

sbidoul commented on contract/models/contract.py:
```diff
@@ -0,0 +1,223 @@
+        for contract in self:
+            contract.date_end = False
+            date_end = contract.recurring_invoice_line_ids.mapped('date_end')
+            if date_end and all(date_end):
```
```suggestion
            if all(date_end):
```

sbejaoui replied:
`all([])` return True, so we should check if date_end is not empty and don't contain False

sbidoul commented on contract/models/contract.py:
```diff
@@ -0,0 +1,223 @@
+                'create_invoice_visibility'
+            ).mapped('recurring_next_date')
+            if recurring_next_date:
+                contract.recurring_next_date = min(recurring_next_date)
```
else contract.recurring_next_date = False?

sbejaoui replied:
I think it's not necessary,

NL66278 replied:
I think it is needed. What if contract.recurring_next_date had a value and user cancels all lines? recurring_next_date variable will be empty, therefore contract will not be updated.

sbidoul commented on contract/models/contract.py:
```diff
@@ -0,0 +1,223 @@
+            contract.date_end = False
+            date_end = contract.recurring_invoice_line_ids.mapped('date_end')
+            if date_end and all(date_end):
+                contract.date_end = max(date_end)
```
else contract.date_end = False?

sbejaoui replied:
computed field are set to false by the model before even the call of the compute method.

sbidoul replied:
Are you 100% sure of that?

sbejaoui replied:
I know that by the use, but not 100% sure

sbejaoui replied:
https://github.com/odoo/odoo/blob/12.0/odoo/fields.py#L1030

sbidoul replied:
ok, thanks!

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        string='Analytic Account',
+        required=True,
+        ondelete='cascade',
+        oldname='analytic_account_id',
```
index = True

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                    )
+                if not rec.date_end:
+                    raise ValidationError(
+                        _("An auto-renew line should have a " "date end ")
```
```suggestion
                        _("An auto-renew line must have a end date")
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                    raise ValidationError(
+                        _(
+                            "A contract line with a successor "
+                            "should have date end"
```
```suggestion
                            "must have a end date"
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    def _check_overlap_successor(self):
+        for rec in self:
+            if rec.date_end and rec.successor_contract_line_id:
+                if rec.date_end > rec.successor_contract_line_id.date_start:
```
```suggestion
                if rec.date_end >= rec.successor_contract_line_id.date_start:
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    def _check_overlap_predecessor(self):
+        for rec in self:
+            if rec.predecessor_contract_line_id:
+                if rec.date_start < rec.predecessor_contract_line_id.date_end:
```
```suggestion
                if rec.date_start <= rec.predecessor_contract_line_id.date_end:
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        )
+
+    @api.onchange(
+        'is_auto_renew', 'auto_renew_rule_type', 'auto_renew_interval'
```
add `date_start` too?

sbejaoui replied:
Right

sbidoul reviewed (commented):
Next batch. Looks quite good overall.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+            if not line.recurring_next_date:
+                raise ValidationError(
+                    _(
+                        "You must supply a next invoicing date for contract "
```
```suggestion
                        "You must supply a date of next invoice for contract "
```

(for consistency with the field name)

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                if line.date_start > line.recurring_next_date:
+                    raise ValidationError(
+                        _(
+                            "You can't have a next invoicing date before the "
```
```suggestion
                            "You can't have a date of next invoice anterior to the "
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    @api.constrains('recurring_next_date')
+    def _check_recurring_next_date_recurring_invoices(self):
+        for line in self.filtered('contract_id.recurring_invoices'):
+            if not line.recurring_next_date:
```
Don't we set recurring_next_date to False where there is nothing to invoice anymore?

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                                line.recurring_rule_type,
+                                line.recurring_interval,
+                            )
+                        ) <= line.date_end
```
line.create_invoice_visibility = False ?

sbejaoui committed:
[IMP] - store last_date_invoiced on contract_line

Improve CRITERIA_ALLOWED_DICT

[IMP] - code improvement

[IMP] - Use last_date_invoiced to set marker in invoice description

[IMP] - add migration script to init last_day_invoiced and some other improvement

[FIX] - a contract line suspended should start a day after the suspension end

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    def _compute_create_invoice_visibility(self):
+        today = fields.Date.today()
+        for line in self:
+            if line.date_start:
```
date_start is required for recurring contract lines? so this if might be redundant

sbejaoui replied:
User can empty this field and compute will be executed with false as value.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    @api.model
+    def recurring_create_invoice(self, contract=False):
+        domain = []
+        date_ref = fields.Date.today()
```
```suggestion
        date_ref = fields.Date.context_today()
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                # and filter with create_invoice_visibility
+            ]
+        )
+        lines = self.search(domain).filtered('create_invoice_visibility')
```
If I understand well, this is going to return all contract lines with recurring_next_date in the past, and then filter on create_invoice_visibility, so this will get slower and slower as past lines accumulate.

Would it not be much simpler to set recurring_next_date to False when there is nothing more to invoice?

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        date_from = fields.Date.from_string(self.recurring_next_date)
+        date_to = date_from + self.get_relative_delta(
+            self.recurring_rule_type, self.recurring_interval
+        )
```
This is not a good place to compute date_from and date_to, they should be passed as parameter to this function, since they must have been computed elsewhere when deciding what we are invoicing.

The logic that computes the period being invoiced deserves it's how method, with unit tests.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+            lines = self.filtered(lambda l: l.contract_id == contract)
+            invoices |= lines._create_invoice()
+            lines._update_recurring_next_date()
+        return invoices
```
I'm not sure this method will be efficient if we are invoicing many lines.

sbejaoui replied:
I changed it with read_group

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    @api.multi
+    def _create_invoice(self):
+        """
+        :param invoice: If not False add lines to this invoice
```
the param does not exist

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    @api.multi
+    def _update_recurring_next_date(self):
+        for line in self:
+            ref_date = line.recurring_next_date or fields.Date.today()
```
why today?

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+    def _update_recurring_next_date(self):
+        for line in self:
+            ref_date = line.recurring_next_date or fields.Date.today()
+            old_date = fields.Date.from_string(ref_date)
```
from_string not needed in v12?

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                rec.date_end
+                if not rec.date_end
+                else rec.date_end + delay_delta
+            )
```
This is harder to read than a plain old:

```python
if rec.date_end:
    rec.date_end = rec.date_end + delay_delta
```

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+            old_date_start = rec.date_start
+            old_date_end = rec.date_end
+            new_date_start = rec.date_start + delay_delta
+            rec.recurring_next_date = self._compute_first_recurring_next_date(
```
Since this calls _compute_first_recurring_next_date, is it correct that we are not supposed to call delay if the first invoice has been sent?

sbejaoui replied:
yes,

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+        :param delay_delta: delay relative delta
+        :return: delayed contract line
+        """
+        for rec in self:
```
First check that conditions to call delay are satisfied?

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+            raise ValidationError(_('Stop not allowed for this line'))
+        for rec in self:
+            if date_end < rec.date_start:
+                rec.cancel()
```
This cancel() side effect is suprising, why is it here?

sbejaoui replied:
When a contract line is stopped before its start date, I think we should cancel it.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,819 @@
+                    rec.date_end
+                    if rec.date_end and rec.date_end < date_end
+                    else date_end
+                )
```
Maybe a question of taste but I think important business logic like this deserves an explicit if/then/else. Especially since we probably want to avoid write and message_post if the end date does not change.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,842 @@
+            if rec.predecessor_contract_line_id:
+                rec.predecessor_contract_line_id.successor_contract_line_id = (
+                    rec
+                )
```
add `assert not  rec.predecessor_contract_line_id.successor_contract_line_id`

sbejaoui replied:
It's handled by is_un_cancel_allowed

sbidoul replied:
Yes, but it's hidden out of view. The assert here would help as a kind of documentation + safety net to make extra sure we do not overwrite the successor of the predecessor if one has been set in the meantime

rafaelbn commented:
Hi, thank you a lot both.

> So the tags and cost/revenue buttons can be seen on the native analytic account view only

OK. But Pedro know we have a customer which this case is a requsite so we will add it in other module in the fuure

> people managing contracts (usually sales persons)

Well Odoo don't think so has contact are in sales :smile: 

Thank again :+1:

sbejaoui committed:
[IMP] - don't allow to unlink uncnaceled contrac line

sbejaoui committed:
[FIX] - check date_start before onchange

sbejaoui committed:
[FIX] - compute recurring_next_date for contract

[IMP] - get contract line default data onchange product_id

sbejaoui committed:
[IMP] - Add responsible to contract form view

sbejaoui committed:
[FIX] - contract recurring_next_date ignore canceled lines

[FIX] - fix _get_invoiced_period if recurring_next_date manually updated

sbejaoui committed:
[IMP] - archive contract_line on contract archive

ThomasBinsfeld committed:
[REF] Contract: invoice creation

[FIX] - Fix typo

[IMP] - date start required for contract line

[REF] Gitignore: .eggs

ThomasBinsfeld committed:
[REF] Contract: invoice creation

[REF] Contract Unit Tests: base the cron check on invoice lines instead of invoices

sbejaoui committed:
[FIX] - as date_start is required, constraints on it become useless

[FIX] - for finished contract _get_lines_to_invoice should return False

sbejaoui committed:
[FIX] - default value for active field in contract line

[FIX] - fix flake8

ThomasBinsfeld committed:
[ADD] Contract: payment term

sbejaoui committed:
[12.0][IMP] - check invoice line vals before assignment

[FIX] - Fix unit tests.

sbejaoui committed:
[FIX] - do not copy last_date_invoiced on plan_successor

sbejaoui committed:
[FIX] - renew only recurring_invoices contract

sbejaoui committed:
[FIX] - filter by termination_notice for contract line to renew

sbejaoui committed:
[IMP] - Improve inprogress search filter

sbejaoui committed:
[IMP] - Link invoice line to contract lines

[IMP] - Add index on contract line

[IMP] - Add is_suspended flag and _search_state

[IMP] - Add is_suspended flag to stop contract line

rafaelbn commented:
Thank! I know  codecov/patch — 87% of diff hit (target 97.6%)  it is not mandatory but I you think is affordable to increase coverage it could be great

Regards

pedrobaeza reviewed (changes requested):
I want to review it as it has a lot of changes that impact in our flows. Any way, test coverage has dropped a lot, so in this case it's blocking as I was keeping this to a sane 100% coverage before.

sbejaoui committed:
[MV] - move onchange method to contract_product module

[RMV] - Remove useless check

sbejaoui committed:
[IMP] - Replace is_suspended flag by manual_renew_needed

Add a computed field for the first date of the termination notice period

Adapt state compute and search method

[IMP] - Improve unit tests

ThomasBinsfeld committed:
[FIX] Contract: copy payment term on generated invoice

sbejaoui committed:
[IMP] - Should be able to stop a past contract line

ThomasBinsfeld committed:
[ADD] Contract: fiscal position

[ADD] Contract: new contributor

[REF] Contract: indexes

sbejaoui committed:
[FIX] - recurring_next_date is required if date_end is null

sbejaoui committed:
[IMP] - rename recurring_interval to Invoice Every

ThomasBinsfeld committed:
[IMP] Contract: matching invoice view depending on contract type

[ADD] Contract: invoice partner

sbejaoui committed:
[ADD] - Add more value for auto_renew_rule_type

[REF] - refactoring for compute_first_date_end

[RMV] - remove auto_renew_rule_type option monthlylastday

[IMP] - rename manual_renew_needed label

[IMP] - use get_product_multiline_description_sale to get contract line description

ThomasBinsfeld committed:
[FIX] Contract to Invoice: take contract vendor instead of partner's

sbejaoui committed:
[FIX] - manual_renew_needed should be visible to all contract users

NL66278 commented on contract/models/abstract_contract.py:
```diff
@@ -0,0 +1,74 @@
+from odoo import api, models, fields
+
+
+class AbstractAccountAnalyticContract(models.AbstractModel):
```
Why the incredibly long name for class and model? I would prefer just AbstractContract and 'abstract.contract' (and the same goes for line). This would also give full consistency between python filename, model name and class name.

sbejaoui replied:
Agree,

sbejaoui replied:
@sbidoul summarized why we are in this "just between" situation

NL66278 commented:
With this large reorganization, that is very welcome and addresses a lot of issues that I encountered with customers, why not go the extra mile and do away with some unfortunate designs that are inherited from the past:
1. Make contracts completely independent from analytic accounts (and name the model just contract);
2. Rename account.analytic.invoice.line to contract.line. I found it very confusing that these invoice lines are not actually invoice lines...;

So we would have:
- abstract.contract (.line)
- contract.template (.line)
- contract (.line)

3. Then give some fields shorter (and less confusing) names:
recurring_invoice_line_ids ==> contract_line_ids

sbidoul commented:
@NL66278 thanks for the review. Glad you like it.

We totally agree with your last mile suggestions. We started the split from analytic account by disconnecting the views already. We did not go further to have a first step that is easy on migration: we were a bit fearing the PR would be rejected if we went too far.

We are currently 1 week from go live on a project that generates 10 000 invoices per month with this module so we can't really change much right now.

Our idea is to have this one merged as 12.0.2.0.0, then we can help/work on a 3.0.0 (on 12 or 13) that does the full split from analytic account and the renaming you mention.

(BTW, @sbejaoui is currently working on a contract_forecast module that uses queue_job to maintain a list of future invoice lines each time a contract line change, for reporting purpose)

pedrobaeza commented:
:+1: for going with total isolation, but we need the migration scripts. Ping me for any doubt in how to do it.

sbejaoui committed:
[REF] - add _get_next_invoiced_period method

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -512,31 +512,57 @@ def _prepare_invoice_line(self, invoice_id=False):
+            self.recurring_next_date,
+            self.recurring_interval,
+            self.recurring_rule_type,
+            self.recurring_invoicing_type,
```
@sbejaoui for the purpose of reusing in the forecasting algorihtm it would be slightly easier to read if _get_next_invoiced_period remains a multi/ensure_one method, taking only last_date_invoiced and recurring_next_date as arguments.

sbejaoui replied:
yes

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -513,30 +513,44 @@ def _prepare_invoice_line(self, invoice_id=False):
+            self.last_date_invoiced, self.recurring_next_date
+        )
+        first_date_invoiced, last_date_invoiced, recurring_next_date = dates
         return first_date_invoiced, last_date_invoiced
```
@sbejaoui another question: it's looks strange that `recurring_next_date` is computed by `_get_next_invoiced_period` then ignored in the return here, and computed (differently) elsewhere in `_update_recurring_next_date`. Also this `_get_invoiced_period` now looks a little bit useless. 

Perhaps you can push the refactoring further?

sbejaoui replied:
yes, I was in a hurry to get out `_get_next_invoiced_period` to use it for the forecast algorithm. Unify the invoiced period compute is in my to do list.

sbejaoui committed:
[FIX] - Fix In progress contract filter

sbejaoui committed:
[REF] - add method to get quantity to invoice

the contract line quantity can be in some use cases variable, to simplify
the way it is changed we add a new method _get_quantity_to_invoice.

sbejaoui committed:
[FIX] - Onchange contract line don't reste contract lines

[FIX] - Rename _get_invoiced_period to _get_period_to_invoice

[ADD] - Add stop_at_date_end to _get_period_to_invoice

[FIX] - Remove useless filter

sbejaoui committed:
[FIX] - don't play onchange date_start for old lines on contract template change

[FIX] - Fix stop post message

[FIX] - Fix sale_contract_count should count all partner contract

[FIX] - set recurring_next_date to False if contract line stoped at last date invoiced

[FIX] - Group by next_invoice also considers dates in the past

MiquelRForgeFlow commented on contract/models/res_partner.py:
```diff
@@ -8,35 +8,47 @@ class ResPartner(models.Model):
-                r[2] for r in result
-                if r[0] in partner_child_ids and r[1] == 'purchase'])
+            partner.sale_contract_count = sum(
+                [
```
I don't understand why you need to do this kind of cosmetics changes. This creates more difference, adds difficulty to trace the originator of these lines and there is not really any visual improvement.

sbejaoui replied:
Sorry about that, I delegate code formatting to [balck](https://github.com/ambv/black) and it's automatically make this "cosmetics changes". I will try to perform a revert on this kind of changes to facilitate code review.

sbejaoui committed:
[IMP] - A canceled contract line can't be set to auto-renew

[REF] Contract: pep8

[IMP] Contract: update USAGE section of README

[IMP]call onchange_is_auto_renew when changing template

rafaelbn reviewed (commented):
@ThomasBinsfeld this improvements it changing the functionality so the Usage section of the readme should be updated. If not new reviewer don't know how to use it.

Also travis is :red_circle: , please could you check it?

Thanks a lot!

ThomasBinsfeld commented:
@rafaelbn Indeed. I'll make Travis green and coordinate with @sbejaoui to update the USAGE section.

AaronHForgeFlow commented:
IMO  the end date for the contract line should be computed when the line es created after onchange_template_id is called. Now the user has to manually add the date in the line even if the renewal period is set in the template:

![image](https://user-images.githubusercontent.com/19620251/58415252-26668b80-807e-11e9-895c-ff53170bf066.png)

@sbejaoui if you agree I can propose a PR to your branch

ThomasBinsfeld commented:
@rafaelbn I updated the USAGE section and Travis should be green now. I'll start the split from analytic account.

sbejaoui commented:
> IMO the end date for the contract line should be computed when the line es created after onchange_template_id is called. Now the user has to manually add the date in the line even if the renewal period is set in the template:
> 
> ![image](https://user-images.githubusercontent.com/19620251/58415252-26668b80-807e-11e9-895c-ff53170bf066.png)
> 
> @sbejaoui if you agree I can propose a PR to your branch

I am not sure. As there is the use case of contract without predefined end date. auto_renew_rule_type is considered only if auto_renew is checked.

AaronHForgeFlow commented:
> I am not sure. As there is the use case of contract without predefined end date. auto_renew_rule_type is considered only if auto_renew is checked.

Yes I noticed that. What I can do is only to fill the end date when the auto_renew is set in the template. Is that fine?

sbejaoui commented:
> > I am not sure. As there is the use case of contract without predefined end date. auto_renew_rule_type is considered only if auto_renew is checked.
> 
> Yes I noticed that. What I can do is only to fill the end date when the auto_renew is set in the template. Is that fine?

I think it's the case with this onchange 

https://github.com/OCA/contract/blob/16c2cd75cfdc2e53902658bb904fa9ff62a05e32/contract/models/contract_line.py#L380-L396

AaronHForgeFlow commented:
> I think it's the case with this onchange

Thanks. https://github.com/sbejaoui/contract/pull/11

AaronHForgeFlow commented:
It seems there are conflicts in files

 ```
contract/i18n/de.po
contract/i18n/gl.po
```
@sbejaoui if you don't mind I can resolve them here.

sbejaoui commented:
> It seems there are conflicts in files
> 
> ```
> contract/i18n/de.po
> contract/i18n/gl.po
> ```
> 
> @sbejaoui if you don't mind I can resolve them here.

yes sure

rafaelbn commented:
@pedrobaeza please review to unblock this PR please, It's quite near to be merged

MiquelRForgeFlow commented:
Could you wait until https://github.com/OCA/contract/pull/337 gets merged?

Also, please do a general rebase and squash, to reduce number of commits.

carlosdauden commented on contract/models/res_partner.py:
```diff
@@ -8,35 +8,43 @@ class ResPartner(models.Model):
     sale_contract_count = fields.Integer(
-        string='Sale Contracts',
-        compute='_compute_contract_count',
+        string='Sale Contracts', compute='_compute_contract_count'
```
Why this changes?

carlosdauden commented on contract/views/contract.xml:
```diff
@@ -0,0 +1,316 @@
+                <separator/>
+                <filter name="not_finished"
+                        string="In progress"
+                        domain="['|', ('date_end', '&gt;=', time.strftime('%Y-%m-%d')), ('date_end', '=', False)]"
```
```suggestion
                        domain="['|', ('date_end', '&gt;=', context_today().strftime('%Y-%m-%d')), ('date_end', '=', False)]"
```

carlosdauden commented on contract/views/contract.xml:
```diff
@@ -0,0 +1,316 @@
+                />
+                <filter name="finished"
+                        string="Finished"
+                        domain="[('date_end', '&lt;', time.strftime('%Y-%m-%d')), ('recurring_next_date', '=', False)]"
```
```suggestion
                        domain="[('date_end', '&lt;', context_today().strftime('%Y-%m-%d')), ('recurring_next_date', '=', False)]"
```

AaronHForgeFlow reviewed (approved):
:+1: Functional tests

AaronHForgeFlow commented:
There is gonna be a minor conflict on the manifest version when merging #337  but it will be easy to solve, this module version number should be taken in this case

AaronHForgeFlow committed:
[IMP]pass the payment term to the contract from the partner

AaronHForgeFlow commented:
The payment terms are not passed from the partner to the contract. I created a PR for that https://github.com/sbejaoui/contract/pull/13

nikul-serpentcs commented on contract/models/contract.py:
```diff
@@ -188,6 +188,12 @@ def _onchange_contract_template_id(self):
     def _onchange_partner_id(self):
         self.pricelist_id = self.partner_id.property_product_pricelist.id
         self.fiscal_position_id = self.partner_id.property_account_position_id
+        if self.contract_type == 'purchase':
```
@sbejaoui No Need `Else`
``
self.payment_term_id = \
                self.partner_id.property_payment_term_id
 if self.contract_type == 'purchase':
self.payment_term_id = \
                self.partner_id.property_supplier_payment_term_id
```

ThomasBinsfeld committed:
[REF] Contract: split from analytic account

ThomasBinsfeld committed:
[REF] Contract Sale: split from analytic account

[REF] Contract Sale: update translations

rafaelbn commented:
@ThomasBinsfeld could you please check the code suggestion of @carlosdauden please?

rafaelbn commented:
@ThomasBinsfeld @sbidoul about 

> "End date on sale order line should recompute when start date changes etc?"

In the google doc. What do you think?

ThomasBinsfeld committed:
[REF] Contract: use context_today instead of time in filters

ThomasBinsfeld commented:
> @ThomasBinsfeld could you please check the code suggestion of @carlosdauden please?

Done

ThomasBinsfeld commented:
@pedrobaeza @rafaelbn @carlosdauden @aheficent @NL66278 @nikul-serpentcs @mreficent @gva-acsone I pushed the split from analytic account with migration scripts. The PR is now ready for a final review.

sbejaoui reviewed (commented):
Thank you @ThomasBinsfeld , good work.
Code review for split commits, LGTM.

sbejaoui commented on contract/models/contract.py:
```diff
@@ -0,0 +1,469 @@
+        inverse_name='contract_id',
+        copy=True,
+    )
+    recurring_invoices = fields.Boolean(
```
@ThomasBinsfeld , this field is not needed anymore

sbejaoui commented on contract/views/contract.xml:
```diff
@@ -0,0 +1,308 @@
+                <field name="journal_id"/>
+                <field name="pricelist_id"/>
+                <separator/>
+                <filter name="recurring_invoices"
```
This filter is not needed anymore

sbejaoui commented on contract/models/contract_line.py:
```diff
@@ -511,12 +514,10 @@ def _prepare_invoice_line(self, invoice_id=False):
         invoice_line_vals.update(
             {
                 'name': name,
-                'account_analytic_id': contract.id,
```
`'account_analytic_id': contract.analytic_account_id.id` ?

ThomasBinsfeld replied:
Analytic accounts on contract lines copied on the invoice lines is a future feature.

sbejaoui replied:
so we will have analytic account in contract line level?

sbejaoui commented on contract/models/contract.py:
```diff
@@ -11,19 +11,36 @@
-        string='Contract Template', comodel_name='account.analytic.contract'
+        string='Contract Template', comodel_name='contract.template'
     )
     recurring_invoice_line_ids = fields.One2many(
```
what do you think if we rename this field to contract_line_ids?

ThomasBinsfeld replied:
Good idea, this name is confusing.

sbejaoui commented on contract/models/contract_template.py:
```diff
@@ -9,13 +9,13 @@
+        comodel_name='contract.template.line',
         inverse_name='contract_id',
         copy=True,
         string='Invoice Lines',
```
`string='Contract Template lines'` ?

sbejaoui commented on contract/views/contract_template_line.xml:
```diff
@@ -1,17 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
I think you removed this file unintentionally.

ThomasBinsfeld replied:
No, in my understanding, it was useless.

sbejaoui replied:
Without this inheritance, contract template lines have no default form view.
If you take a look at runbot, when I add a line to a contract template, I get this view with all model fields.
Impossibility to save as contract template is required.
![1](https://user-images.githubusercontent.com/12665409/58544631-ee885100-8201-11e9-87de-bf96b6948b2e.png)

ThomasBinsfeld replied:
Ok! Thanks.

sbejaoui commented on contract/models/contract.py:
```diff
@@ -11,19 +11,36 @@
+        string='Contract Template', comodel_name='contract.template'
     )
     recurring_invoice_line_ids = fields.One2many(
         string='Invoice Lines',
```
`string='Contract Lines',` ?

xavierjimenez reviewed (approved):
Have checked it functionally and everything seems to be working fine, great work @ThomasBinsfeld !

AaronHForgeFlow commented on contract/migrations/12.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,24 @@
+    set recurring_next_date to false for finished contract
+    """
+    _logger.info("order all contract line")
+    with api.Environment(cr, SUPERUSER_ID, {}) as env:
```
I am getting an error when running the migration here: 

AttributeError: __exit__

But I am not sure what is wrong

sbejaoui replied:
Hi @aheficent ,
I pushed a new version of the migration scripts. I think it resolve this issue

pedrobaeza commented on contract/__manifest__.py:
```diff
@@ -4,29 +4,36 @@
-              "Tecnativa, "
-              "LasLabs, "
-              "Odoo Community Association (OCA)",
+    "Tecnativa, "
```
Please don't de-indent this

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,145 @@
+    )
+    cr.execute(
+        """
+        INSERT INTO contract_line (
```
No need to do this, but only renaming the table and the link field column (in pre).

ThomasBinsfeld replied:
I couldn't do a renaming of the contract lines table in the pre-migration script because of the foreign key on contracts (which can't be renamed since analytic accounts must be preserved). So I have to migrate contracts and their lines the same way.

sbejaoui replied:
@pedrobaeza,
As @ThomasBinsfeld mentioned, we can not rename the contract line table before we prepare the contract table.
How can we solve this problem?
I was looking for a way to create the contract table in the pre-migration script and insert the data there. Thus, the change of contract line name became possible.
I did not find an easy way to do it. In the end, it seems to me that the way things are done now is easy to understand if there is no decent alternative.

pedrobaeza commented on contract/migrations/12.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,24 @@
+_logger = logging.getLogger(__name__)
+
+
+def migrate(cr, version):
```
Please use openupgradelib and decorate it with `@openupgrade.migrate()`. This way, you get a `env` proper populated.

You should also use `openupgrade.logged_query` method for registering the query and the result in the log, which is valuable.

ThomasBinsfeld committed:
[REF] Contract Sale Invoicing: split from analytic account

[REF] Contract Sale Invoicing: update translations

[IMP] - Assert that the predecessor is available for new link at uncancel

[RMV] - remove usless changes

[RMV] - Remove usless field recurring_invoices

after the total isolation between contract model and account analytic one.
recurring_invoices which was used to mark analytic account as contract became usless

[IMP] - P3 syntax

[IMP] - use @openupgrade.migrate() and openupgrade.logged_query

[IMP] - drop transient table in migration script

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,145 @@
+        FROM account_analytic_invoice_line
+        """
+    )
+    openupgrade.rename_models(cr, [('account.analytic.invoice.line',
```
Call also `rename_tables` and you don't need the insert SQL.

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,145 @@
+                                    'contract.line')])
+    cr.execute(
+        """
+        DROP TABLE account_analytic_invoice_line
```
No need to do that with the renaming

pedrobaeza replied:
Never drop any data, and it fails due to constraints and foreign keys.

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,145 @@
+               write_uid,
+               write_date
+        FROM account_analytic_account
+        WHERE recurring_invoices = TRUE
```
You can remove `= TRUE` part

pedrobaeza commented on contract/migrations/12.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,24 @@
+    _logger.info("order all contract line")
+    with api.Environment(cr, SUPERUSER_ID, {}) as env:
+        contracts = env["account.analytic.account"].search([])
+        finished_contract = contracts.filtered(
```
You can't do this on pre. See my alternative solution

pedrobaeza commented on contract/migrations/12.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,24 @@
+        )
+        cr.execute(
+            "UPDATE account_analytic_account set recurring_next_date=null "
+            "where id in (%)" % ','.join(finished_contract.ids)
```
Replace this by `WHERE date_end IS NULL OR recurring_next_date <= date_end`

pedrobaeza commented on contract/models/contract.py:
```diff
@@ -0,0 +1,470 @@
+        ondelete='restrict',
+    )
+    partner_id = fields.Many2one(
+        comodel_name='res.partner', inverse='_inverse_partner_id'
```
This is not a computed field. Why do you use an inverse?

ThomasBinsfeld replied:
A field doesn't have to be computed to use an inverse on it. ping @sbejaoui

pedrobaeza replied:
Isn't this weird? For what is this intended?

sbejaoui replied:
IMO, we can remove the inverse method as `_inverse_partner_id` code is redundant with `_onchange_partner_id`

pedrobaeza replied:
Better to keep inverse method, and on onchange, call to the inverse method. This way, you cover both programatical and UI update.

pedrobaeza commented on contract/migrations/12.0.4.0.0/pre-migration.py:
```diff
@@ -0,0 +1,72 @@
+    openupgrade.rename_models(cr, models_to_rename)
+    openupgrade.rename_tables(cr, tables_to_rename)
+    openupgrade.rename_xmlids(cr, xmlids_to_rename)
+    # A temporary column is needed to avoid breaking the foreign key constraint
```
This shouldn't be needed if you only rename the column

pedrobaeza commented on contract/migrations/12.0.2.0.0/post-migration.py:
```diff
@@ -0,0 +1,32 @@
+
+    _logger.info("order all contract line")
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    contract_lines = env["account.analytic.invoice.line"].search(
```
This will raise an error on current code state, as the model `account.analytic.invoice.line` is not existing anymore.

pedrobaeza commented on contract/migrations/12.0.2.0.0/post-migration.py:
```diff
@@ -0,0 +1,32 @@
+        recurring_next_date=contract.recurring_next_date,
+        date_start=contract.date_start,
+        date_end=contract.date_end
+        FROM account_analytic_account AS contract
```
As renaming is done in pre, you have to put here new table names

pedrobaeza reviewed (changes requested):
First review on migration scripts

pedrobaeza commented on contract_sale_invoicing/models/contract.py:
```diff
@@ -28,5 +28,5 @@ def _create_invoice(self, invoice=False):
             invoice_ids = sales.action_invoice_create()
             invoice = self.env['account.invoice'].browse(invoice_ids)[:1]
-        return super(AccountAnalyticAccount, self)._create_invoice(
+        return super(ContractContract, self)._create_invoice(
```
You can take the occasion to update this to P3 syntax and avoid the class name.

pedrobaeza commented on contract_sale_invoicing/models/contract.py:
```diff
@@ -16,9 +16,9 @@ class AccountAnalyticAccount(models.Model):
     def _create_invoice(self, invoice=False):
         if not self.invoicing_sales:
-            return super(AccountAnalyticAccount, self)._create_invoice()
+            return super(ContractContract, self)._create_invoice()
```
You can take the occasion to update this to P3 syntax and avoid the class name.

pedrobaeza commented on contract/tests/test_contract.py:
```diff
@@ -1,83 +1,129 @@
 
 
+def to_date(date):
+    return fields.Date.to_date(date)
```
For not needing to define this method:
```python
from fields.Date import to_date
```

ThomasBinsfeld replied:
I'm afraid we can't to that.

`ModuleNotFoundError: No module named 'odoo.fields.Date'; 'odoo.fields' is not a package`

pedrobaeza commented on contract/models/res_partner.py:
```diff
@@ -50,19 +58,20 @@ def act_show_contract(self):
             context=dict(
                 self.env.context,
                 search_default_recurring_invoices=True,
```
Remove this key when removing the filter

pedrobaeza reviewed (commented):
Some more comments.

You need also to manually updated some noupdate fields like cron record. I have detected these 3 at least:

- contract_cron_for_invoice
- email_contract_template
- rule_contract_template_multi_company

you can use `openupgrade.load_data` method for that.

pedrobaeza commented:
Can you rebase and solve conflicts after merging the minor change and what are the expected date for the changes we talked in OCA Spanish days?

rafaelbn commented:
@ThomasBinsfeld @sbidoul This branch has conflicts that must be resolved. After that could we merge? or you plan something else? thanks!

pedrobaeza commented:
Any news on this?

ThomasBinsfeld commented:
@pedrobaeza I should be able to finalize it this week.

pedrobaeza commented:
OK, thanks. Let me know when finished.

pedrobaeza commented:
Any news on this?

sbejaoui committed:
[IMP] - Make one migration from 12.0.1 to 12.0.4

[RMV] - Remove empty file

[IMP] - Update no_update cron after migration

[IMP] - move contract template recurrence info to line level

[FIX] - Fix contract line model description

[IMP] - Link contracts to analytic accounts

[FIX] - Fix pylint

[IMP] - Move chatter and attachments from analytic account to contract

[IMP] - Move account_analytic_id to contract line level

[IMP] - Improve version check in migration script

[IMP] - Move contracts followers from analytic accounts

[ADD] - Add mail.activity.mixin to contract.contract model

remove data drop from migration scripts

[12.0][FIX] - Fix _init_last_date_invoiced

fix flake8

[ADD] - Update contributors list

sbejaoui commented:
Hi,
I pushed a new version of the migration script:

- Move recurrence info from contract template to contract template lines
- Keep link between the contract lines and original analytic account
- Get rid of recurring_invoices filed on the contract
- Update no_update data after migration
- Move chatter and attachments from analytic account to contract

I did a lot of database migration tests from 12.0.1.0.0 to 12.0.4.0.0 and everything seems working fine.

@pedrobaeza , can review please.

pedrobaeza commented:
The main problem I see now is that you have migration scripts for 2 version (12.0.2.0.0 and 12.0.4.0.0). Are they compatible for being used in a v11 database?

sbejaoui commented:
> The main problem I see now is that you have migration scripts for 2 version (12.0.2.0.0 and 12.0.4.0.0). Are they compatible for being used in a v11 database?

Yea, I didn't totally think about that.
I made some checks on current version as we are using this PR in production and our current version is 12.0.3.0.0

But I believe that we can move the 12.0.2.0.0 into the 12.0.4.0.0. 
I will see for our installation if I keep the check on version number or made a patch for it.

sbejaoui commented:
> The main problem I see now is that you have migration scripts for 2 version (12.0.2.0.0 and 12.0.4.0.0). Are they compatible for being used in a v11 database?

I improved the way I check the current version number.
https://github.com/OCA/contract/blob/526860425e46ae18aae65f88e9d8b5532da42041/contract/migrations/12.0.4.0.0/post-migration.py#L213-L216

The 12.0.2.0.0 no longer contains post-migration script and the pre-migration uses sql query only.
So, it can be used in a v11 database

sbidoul commented:
@pedrobaeza I think all comments have been handled here.

rafaelbn commented:
This PRs is a historic one 😄

rafaelbn commented:
Please @sbejaoui could you rebsase please?

This branch has conflicts that must be resolved

Thanks!

sbidoul commented:
rebased

rafaelbn commented:
Please @sbidoul could you rebase again ? The are PR arround v12 branch that must take care of this PR and the could want to test this. 

By the way we should merge this :smile: deadline next OCA code sprint!

sbejaoui commented:
> Please @sbidoul could you rebase again ? The are PR arround v12 branch that must take care of this PR and the could want to test this.
> 
> By the way we should merge this  deadline next OCA code sprint!

Hi @rafaelbn ,
Rebase done. 
Appreciate your efforts.
We must take care of this one ASAP to resolve other pending PRs depending on it.

pedrobaeza commented:
Yes, I have been able to almost migrate our own database to v12 for trying this whole PR, as we use intensively contracts and thus we can test everything. I expect to be able to do it next week.

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,229 @@
+            cr,
+            """
+            ALTER TABLE account_invoice
+            DROP COLUMN old_contract_id_tmp
```
Never drop any data

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,229 @@
+        cr,
+        """
+        ALTER TABLE account_invoice_line
+        DROP COLUMN contract_line_id_tmp
```
Never drop any data. No matter is a temporary column you have created. This data can be needed for other dependent modules that needs the link for complementing some data.

pedrobaeza commented on contract/migrations/12.0.4.0.0/pre-migration.py:
```diff
@@ -0,0 +1,118 @@
+    openupgrade.logged_query(
+        cr,
+        """
+        DROP TABLE IF EXISTS account_analytic_invoice_line_wizard
```
Never drop any data

pedrobaeza reviewed (changes requested):
@sbejaoui I'm trying a migration and due to the droppings, there are errors. I can't push to your branch. Have you removed this ability? And you have a lot of commits that doesn't add too much value. Can you squash a bit commit history?

pedrobaeza commented:
I'm finding more problems, now:
```
  File "/opt/odoo/auto/addons/contract/migrations/12.0.4.0.0/post-migration.py", line 187, in _init_last_date_invoiced_on_contract_lines
    contract_lines._init_last_date_invoiced()
  File "/opt/odoo/auto/addons/contract/models/contract_line.py", line 619, in _init_last_date_invoiced
    if last_date_invoiced > rec.date_start:
UnboundLocalError: local variable 'last_date_invoiced' referenced before assignment
```
I have created a temp branch in my fork until I adjust everything and tell you back.

sbejaoui commented:
@pedrobaeza , you can push to my branch now

pedrobaeza commented:
OK, thanks! Can I squash commits by logic set/author?

Any way, I'm going to continue pushing to my local branch until the migration finishes for not hammering this PR.

sbejaoui commented:
> I'm finding more problems, now:
> 
> ```
>   File "/opt/odoo/auto/addons/contract/migrations/12.0.4.0.0/post-migration.py", line 187, in _init_last_date_invoiced_on_contract_lines
>     contract_lines._init_last_date_invoiced()
>   File "/opt/odoo/auto/addons/contract/models/contract_line.py", line 619, in _init_last_date_invoiced
>     if last_date_invoiced > rec.date_start:
> UnboundLocalError: local variable 'last_date_invoiced' referenced before assignment
> ```
> 
> I have created a temp branch in my fork until I adjust everything and tell you back.

I am pushing a fix for this.

pedrobaeza commented on contract/models/contract_line.py:
```diff
@@ -0,0 +1,1105 @@
+                        rec.recurring_rule_type, rec.recurring_interval
+                    )
+                ) - relativedelta(days=1)
+            if last_date_invoiced > rec.date_start:
```
This fails as there can be the option of `rec.recurring_invoicing_type = 'pre-paid'` and `recurring_rule_type != 'monthlylastday'`, but not sure about the strategy to fix it:

- Pre-filter before the records that comply with this.
- Set another `last_date_invoiced` in that cases.

@sbejaoui can you help me?

sbejaoui replied:
I just pushed a fix for it. 
for pre-paid contract, the last date invoiced is the day before the next invoice date.
for the post-paid, it's is the last day of the invoicing period before the next invoice date.

A special treatment for the `monthlylastday` as it's a special case of the post-paid method.

pedrobaeza commented:
> I am pushing a fix for this.

Great, because I have just asked you in a comment about how to fix it

sbejaoui commented:
> OK, thanks! Can I squash commits by logic set/author?
> 
> Any way, I'm going to continue pushing to my local branch until the migration finishes for not hammering this PR.

Feel free please to push directly to the PR please. But let squash commits when you finish your review.

rafaelbn commented on contract/readme/CONTRIBUTORS.rst:
```diff
@@ -5,3 +5,5 @@
 * Vicent Cubells <vicent.cubells@tecnativa.com>
 * Miquel Raïch <miquel.raich@eficent.com>
 * Souheil Bejaoui <souheil.bejaoui@acsone.eu>
+* Thomas Binsfeld <thomas.binsfeld@acsone.eu>
```
Hello, I should be as a contributor IMHO from long time ago (as least from v9). If you agree, please add me.

sbejaoui replied:
yes sure.

rafaelbn reviewed (dismissed):
I saw 2 things:

1) In the contract the button "Create invoice" is only show when debug mode is activated. I think i should be visible all the time.

![2019-09-13_2-00-52](https://user-images.githubusercontent.com/8736623/64829091-2c4f9080-d5cb-11e9-808d-183c9acbb126.png)

2) The field reference (`code`) in contract should be populated to the field reference (`name`) in the invoice and not in the field "Payment ref." (`reference`), which it was used in v11 and before only for Vendors and now in v12 for a payment "Sequence" usualy really similar to the invoice number

![2019-09-13_1-58-07](https://user-images.githubusercontent.com/8736623/64829197-ba2b7b80-d5cb-11e9-8598-40235b047bdf.png)

sbejaoui commented:
> I saw 2 things:
> 
> 1. In the contract the button "Create invoice" is only show when debug mode is activated. I think i should be visible all the time.
> 

It was already the case in all the previous versions and I think it's correct from an UX point of view.
IMO, the manual creation of invoices is not the normal case scenario and invoices should be created by the planned action. That action is meant to be used for debugging reasons.

But we can imagine a new security group "Can manually create invoices from contracts" to restrict the visibility of the button instead of using the debug mode. 

But still for me, the debug mode make sense to hide/show the create invoice button.

> ![2019-09-13_2-00-52](https://user-images.githubusercontent.com/8736623/64829091-2c4f9080-d5cb-11e9-808d-183c9acbb126.png)
> 
> 1. The field reference (`code`) in contract should be populated to the field reference (`name`) in the invoice and not in the field "Payment ref." (`reference`), which it was used in v11 and before only for Vendors and now in v12 for a payment "Sequence" usualy really similar to the invoice number
> 
> ![2019-09-13_1-58-07](https://user-images.githubusercontent.com/8736623/64829197-ba2b7b80-d5cb-11e9-8598-40235b047bdf.png)

I get your point and it make sense for me. But it's the case in [12.0](https://github.com/OCA/contract/blob/12.0/contract/models/account_analytic_account.py#L232). Can you create an issue for this.

pedrobaeza committed:
[FIX+IMP+MIG] contract: Several refinements:

* Remove incorrect oldname attributes.
* Add filter on partners for running contracts (+ a support o2m field for that).
* Cover more tables in model renaming + cleaner code using a loop.
* Don't copy contract lines, but rename table + copy contract records on pre.
* Contract code is now populated to "Reference/Description" field in invoice.
* Order on new contract model has been restored to the same as old analytic accounts.

pedrobaeza commented on contract/migrations/12.0.4.0.0/post-migration.py:
```diff
@@ -0,0 +1,210 @@
+
+    openupgrade.logged_query(
+        cr,
+        "INSERT INTO contract_line ("
```
@sbejaoui  what is the reason for inserting records instead of renaming the table? I'm just experimenting now with the renaming to see problems, but I don't see in theory anyone.

sbejaoui replied:
As we are not renaming the `account_analytic_account` table, renaming the `account_analytic_invoice_line` table in pre-migration trigger foreign-key constraint to `contract_contract` table as it's not created yet.

With @ThomasBinsfeld , we think we had two solutions to resolve this.
1- Copy the create the `contract_contract` and copy in it the `account_analytic_account` table in the pre-migration script. Rename `account_analytic_invoice_line` table to contract_line.
2- Let odoo create the `contract_contract` and the `contract_line` tables and use insert to fill the data.

Do you think we can resolve the foreign-key constraint issue differently?

pedrobaeza replied:
Yes, I think so. Usually the best option is to directly remove that FK constraint in pre and let Odoo to do rebuild it, but I'm trying right now. What are the benefits of not needed to move data?:

1. Speed: if you have a lot of records, insert operation takes time. As this is by SQL, it's not too much, but it's something.
2. Additional modules that added in v11 fields to this table, will need to create extra migration scripts for moving these fields values to the new table with this approach. If we simply rename the table, SQL columns will be there and then no migration script will be needed.
3. Avoid FK constraints.

What I'm going to try is to create table copying columns as is. Having extra columns there (coming from account_analytic_account table) is not problem, as they can be cleaned afterwards.

pedrobaeza commented:
I have finished the check and everything seems OK. A fine work, congratulations!

I have added the following changes, following the previous conversation about not to copy lines, but rename the table (a treacherous oldname was provoking an error without a hint of what was happening):

* Remove incorrect oldname attributes.
* Add filter on partners for running contracts (+ a support o2m field for that).
* Cover more tables in model renaming + cleaner code using a loop.
* Don't copy contract lines, but rename table + copy contract records on pre.
* Contract code is now populated to "Reference/Description" field in invoice.
* Order on new contract model has been restored to the same as old analytic accounts.
* Added Acsone as co-author (you deserve it!).

If you agree, I finally merge this PR.

pedrobaeza commented:
Ah, please squash as commented!

sbejaoui commented:
> Ah, please squash as commented!

Great job, thank you @pedrobaeza for the efforts.
I am squashing now.

sbejaoui commented:
@pedrobaeza , squash is done. PR is ready to merge.

pedrobaeza commented:
/ocabot merge

Hope to see you all on Odoo XP.

rafaelbn commented:
Hello @sbidoul , @sbejaoui !

Did you made this promise changes?

https://github.com/OCA/contract/pull/207#issuecomment-442759393

> We will add the UX feature to manage recurrence globally at the contract level a bit later, don't worry.

I see this is not available in v12 but it is in v13 and v14. And I cannot find where is that improvement in 13 or 14. I only know that it's not in 12 and today that contract module have FIX FIX FIX FIX and stills doesn't work propertly. Futhermore there are already BUGs in 13 and 14. https://www.loom.com/share/86fc343c704c42b6b5ab3c7b86828d29
